### ISSUE TITLE
[mlir] OpInterfaces with verifiers which are dynamically attached don't get verified

### ISSUE THREAD

**Issue by teqdruid** [llvm/llvm-project#64212]
I'm attaching an OpInterface (defined in ODS with a `verify` field) via `attachInterface` (on to `mlir::ModuleOp` if it matters. The attach works (it tests correctly with `isa`) but the verify code never gets run. https://github.com/llvm/circt/pull/5716 is where I'm trying to do this.

Key snippets (simplified):

The attach call, which I've verified runs:
`  mlir::ModuleOp::attachInterface<InnerRefNamespaceOpInterface>(*getContext());`

The ODS for the op interface:
```
def InnerRefNamespace : OpInterface<"InnerRefNamespaceOpInterface"> {
  let cppNamespace = "circt::hw";
  let verify = [{
    llvm::outs() << "Runing IRN verifier on " << op->getName() << "\n";
    return success();
  }];
  // let verifyWithRegions = 1;
}
```

Doesn't work regardless of whether or not `verifyWithRegions` is enabled or not.

Am I doing something wrong or is this bug?

**Comment by teqdruid** [comment]
Sorry @EugeneZelenko typo hitting enter before I was done entering the issue.

**Comment by teqdruid** [comment]
Just to add more data: the verifier runs properly on ops which include the interface statically in their ODS.

**Comment by joker-eph** [comment]
I expect it, the verifier are part of the Trait and not the interface itself. We can dynamically add an operation but not the verifier traits.

It is possible to change all this, but this is pretty deep into the core infra, so it'd be fairly involved to revisit all this. There may be some workaround for this we can support though?

Adding a few folks to help think about it.

**Comment by Mogball** [comment]
@joker-eph is correct: the interface "verifier" actually forms the interface trait verifier that is statically attached to the operation. However, this is indeed confusing for dynamically-attached interfaces, because one would expect operations that implement an interface to adhere to its invariants, including whatever is specified by the verifier.

This is also true of the dependent traits of interfaces: these are verified statically and not dynamically. Perhaps interfaces can get a dynamic verifier hook that is called by the operation verifier.

**Comment by joker-eph** [comment]
Oh the solution is obvious actually and we already have a plan for this: it's the promised interface.
We just rolled the promised DialectInterface but now we need the Op, Attribute, and Type variant.

With the promised interface you **must** declare the interface support on the operation in ODS, even if you provide the implementation dynamically.

**Comment by ftynse** [comment]
I have a vague recollection of @matthias-springer hitting and looking into this issue in his bufferization work. 

> This is also true of the dependent traits of interfaces: these are verified statically and not dynamically. Perhaps interfaces can get a dynamic verifier hook that is called by the operation verifier.

Could we have the top-level verifier iterate over the list of known interfaces for the op and call verifiers from those?

> Oh the solution is obvious actually and we already have a plan for this: it's the promised interface.
We just rolled the promised DialectInterface but now we need the Op, Attribute, and Type variant.
>
> With the promised interface you must declare the interface support on the operation in ODS, even if you provide the implementation dynamically.

This would be problematic for the main use case of dynamic interface registration. We usually resort to it when we don't want the op itself to depend on the interface. The first message in the thread mentions `ModuleOp` and a CIRCT interface. We can't (and arguably don't want to) add a CIRCT interface declaration in `BuiltinOps.td`.

**Comment by matthias-springer** [comment]
> This would be problematic for the main use case of dynamic interface registration. We usually resort to it when we don't want the op itself to depend on the interface. 

There is an example of this even inside MLIR: `BufferizableOpInterface` is currently defined in the bufferization dialect. Ops from other dialects cannot declare this interface as "promised" without depending on the bufferization dialect.

> It is possible to change all this, but this is pretty deep into the core infra, so it'd be fairly involved to revisit all this.

I was trying to do that a few months ago, but gave up because it was getting quite complicated. The verifier could be seen as an `InterfaceMethod`; it just exists multiple times (one per interface). Maybe that helps somehow...

**Comment by teqdruid** [comment]
> This is also true of the dependent traits of interfaces: these are verified statically and not dynamically. Perhaps interfaces can get a dynamic verifier hook that is called by the operation verifier

Yeah, looking through the MLIR code I saw that all verifiers were attached statically. Good to know I didn't miss anything.

> The first message in the thread mentions `ModuleOp` and a CIRCT interface. We can't (and arguably don't want to) add a CIRCT interface declaration in `BuiltinOps.td`.

I agree -- that's not the solution I was looking for. It would be acceptable, however, to have certain upstream classes implement a "DynamicVerifiersOpInterface" and only support dynamically attaching verifiers to those ops. While this would work for my use-case, it doesn't solve the problem of unintuitive behavior for dynamically attached interfaces.

> So it isn't terribly complicated to implement what I had in mind, here's a draft: https://reviews.llvm.org/D156688. Interface verification flow becomes entirely dynamic, but has a bonus effect of also allowing verifiers on type and attribute interfaces.

This is more-or-less what I had expected to see. I like this proposed change.

**Comment by joker-eph** [comment]
> This would be problematic for the main use case of dynamic interface registration. We usually resort to it when we don't want the op itself to depend on the interface. The first message in the thread mentions ModuleOp and a CIRCT interface. We can't (and arguably don't want to) add a CIRCT interface declaration in BuiltinOps.td.

This is a problem with the way we rolled out external interface: this was a big mistake and I regret not pushing back on landing it **without** the promised mechanism first. Now we have to back out of this somehow: this is not conductive to a robust system. Folks should not add interfaces on op that haven't been designed on or that they don't own: if you feel like ModuleOp should expose an interface you should propose one upstream or use your own ModuleOp.

> There is an example of this even inside MLIR: BufferizableOpInterface is currently defined in the bufferization dialect. Ops from other dialects cannot declare this interface as "promised" without depending on the bufferization dialect.

Technically interfaces aren't belonging to any dialect, it's not like type/attributes/operations. For example builtin is able to provide implementation for the bytecode interface which is defined in the bytecode library, without introducing a dependency between the two libs.

**Comment by ftynse** [comment]
> This is a problem with the way we rolled out external interface: this was a big mistake and I regret not pushing back on landing it without the promised mechanism first. Now we have to back out of this somehow: this is not conductive to a robust system. Folks should not add interfaces on op that haven't been designed on or that they don't own: if you feel like ModuleOp should expose an interface you should propose one upstream or use your own ModuleOp.

I believe there was a consensus that adding an interface to an op requires one to "own" either the op or the interface (or both). So in my mind it is perfectly fine define an interface downstream, e.g. "ExecutableOnMyAccelerator", and have it attached to upstream operations in the downstream configuration. The downstream "owns" the interface so it was presumably designed to support the relevant upstream ops. I can see good reasons not to add the interface, or an overgeneralized version of it, both to keep downstream details private and to avoid polluting upstream with things that don't really belong there. If we don't give this possibility, we will be encouraging duplication of all upstream dialects (since it is _singificantly_ cheaper than adding an interface).

**Comment by teqdruid** [comment]
> I believe there was a consensus that adding an interface to an op requires one to "own" either the op or the interface (or both). So in my mind it is perfectly fine define an interface downstream, e.g. "ExecutableOnMyAccelerator", and have it attached to upstream operations in the downstream configuration. The downstream "owns" the interface so it was presumably designed to support the relevant upstream ops.

In this case, one of our dialects "owns" the interface (read "owns" as generates header files in the same directory as) and we'd like to add it to an upstream op.

> I can see good reasons not to add the interface, or an overgeneralized version of it, both to keep downstream details private and to avoid polluting upstream with things that don't really belong there. If we don't give this possibility, we will be encouraging duplication of all upstream dialects (since it is _singificantly_ cheaper than adding an interface).

I agree completely. There are situations (like this one) where ops (or their interfaces/traits) require that they live under an op with a particular interface, but sibling ops expect to live under a different one (e.g. so they share a symbol table). In general, this is the problem with custom top-level ops. Being able to attach interfaces dynamically is a way to solve that problem. This function increases composibility.

**Comment by joker-eph** [comment]
> In this case, one of our dialects "owns" the interface (read "owns" as generates header files in the same directory as) and we'd like to add it to an upstream op.

Right, I sympathize with the use case, but as long as there is no "safe" way to build this into a consistent system: I'd rather just forbid this.
Ans I believe this has been the goal since day1, I don't remember this consensus @ftynse :  in my memory the external interface was added with the agreement that we'd add **quickly** the promised interface feature, this just fell through the crack.

My main concern here is that we're exposing a fragile system: back to "the JSON of compilers", we're losing the ability to check for consistency of the  system, and that breaks reproducibility.
"Why do I see miscompile/crash/suboptimal results?" 
"You forgot to register an external interface..."
"Uh..."

That's just a no-go to me: it's in line with the getDependentDialects and other features we enforced over time.

**Comment by ftynse** [comment]
I can see your point. Let's try to elaborate the situations where the inconsistent behavior may happen.

1. a client of upstream sees a different behavior around upstream ops because they didn't register an upstream interface;
2. a client of CIRCT (using for brevity here, any downstream project works) sees a different behaviour around a CIRCT op because they didn't register a CIRCT interface;
3. a client of CIRCT sees a different behavior around an _upstream_ op because they didn't register an _upstream_ interface;
4. a client of CIRCT sees a different behavior around an _upstream_ op because they didn't register a CIRCT interface for it;
5. a client of CIRCT sees a different behavior around a CIRCT op because they didn't register an _upstream_ interface for it.

1 and 2 are essentially the same if we don't treat upstream specially. A client of an MLIR-enabled project (upstream or CIRCT) needs information that the interface must be registered. Since both the op and the interface belong to the same project, promised interfaces are able to solve this. They should be able to also solve 3, which is a transitive case of 1.

Now, 4 and 5 are more complicated because the op and the interface belong to different projects. Since there is already some sort of dependency of CIRCT on upstream, we could likely have CIRCT specify a promise on its op that it implements an upstream interface (if for some reason it wants to use delayed registration instead of directly specifying the interface on the op), so this addresses 5.

Remains 4. I would very much prefer constraining this to CIRCT and its clients so upstream doesn't need to care about it. This would require a mechanism that allows CIRCT to promise _its clients_ that it implements a CIRCT interface for upstream ops _without_ modifying upstream. Clients of upstream don't care about CIRCT interfaces. If we can figure out the way to do this with some flavor of promised interfaces, it should solve both the consistency and the flexibility issues.

I see two potential approaches:

A. promises can be specified in a dialect extension; this extension is listed in dependencies of a pass that needs the promised interfaces; if by the time the pass runs the promise has not been fulfilled, an error is emitted;
B. promises can be specified in some new project-wide registration hook.

**Comment by joker-eph** [comment]
One concrete instance of the situation I am concerned about, we just had some issue with our setup with a tool calling bufferization, it worked fine with mlir-opt but  failed with `loc("-":1:10): error: op was not bufferized` in the other tool.
Turns out we were missing: `mlir::bufferization::func_ext::registerBufferizableOpInterfaceExternalModels(registry)`.

This is just one example, I'm reporting here to emphasize that we're not talking about "hypothetical situations" here..

> This would require a mechanism that allows CIRCT to promise its clients that it implements a CIRCT interface for upstream ops without modifying upstream. 

I am not totally convinced that interfaces are meant for this: defining that an operation implements an interface seems like part of the dialect definition. You're saying that you may not own a dialect, but want to change the definition. What about traits and other thing about the op? What when the dialect changes?

> A. promises can be specified in a dialect extension; this extension is listed in dependencies of a pass that needs the promised interfaces; if by the time the pass runs the promise has not been fulfilled, an error is emitted;

I don't quite get the part about "this extension is listed in dependencies of a pass"? To me a promise is attached to an operation, how can a pass list a promise?

**Comment by ftynse** [comment]
> This is just one example, I'm reporting here to emphasize that we're not talking about "hypothetical situations" here..

Absolutely! I once spent an entire week fixing our internal projects after a change to the func dialect to use interface... But I haven't necessarily hit all of the combinations, so I'm proactively listing them.

> I am not totally convinced that interfaces are meant for this: defining that an operation implements an interface seems like part of the dialect definition. You're saying that you may not own a dialect, but want to change the definition. What about traits and other thing about the op? What when the dialect changes?

I see another interpretation of interfaces: they "explain" the semantics of the operation in some terms relevant for a particular analysis/transformation, and may or may not be part of the dialect definition. For example, `cf.br` transfers control flow to a different block and potentially forwards values to block arguments. Its implementation of `BranchOpInterface` describes this in terms that dataflow analyses can understand, but it doesn't necessarily mean that `cf.br` transfers control flow _because_ it implements the interface, it does so regardless.

The closest programming analogy I have are typeclasses in Haskell. A type exists on its own, and one can say it is an instance of a particular typeclass by saying how the "methods" of the typeclass are implemented for objects of this type.

Let me use Enzyme as an example of derived project with interfaces because I don't understand CIRCT enough. Enzyme has an `AutoDiffOpInterface` for operations that can be differentiated. It provides implementations of this interface for a bunch of operations from `arith`. These implementations merely interpret arith ops for Enzyme, but don't change their semantics in any way: `arith.mulf` is differentiable because the gradient of a*b exists. Though I don't think we necessarily want to bring that knowledge in form of Enzyme interface into upstream.

Now, if the definition of `arith.mulf` changes so that its semantics no longer matches the autodiff interface implementation, we've got a problem. I would argue it is up to Enzyme to update the interface accordingly in this case. I would further argue that interfaces are not special here: if Enzyme had a canonicalization pattern involving its custom ops and `arith.mulf`, it might have become invalid; basically any logic that assumed certain semantics of the operation might have to be revised. 

There's also a technical annoyance: if I don't have a mechanism to register Enzyme autodiff interfaces on the upstream ops, I'd have to either put a huge switch over the op kind instead or reimplement all those ops within Enzyme (+bidirectional conversion) so I can put the interface on them. Neither sounds like good extensibility.

> I don't quite get the part about "this extension is listed in dependencies of a pass"? To me a promise is attached to an operation, how can a pass list a promise?

The first part says "[if] promises can be specified in a dialect extension". The extension can be registered in `getDependentDialects` of the "differentiate" pass the Enzyme case. It feels like there are ways around this, and one may also just directly register the extension with interface implementations instead.

I also thought about switching the dependency direction and have the interface promise implementations for upstream dialects, but it sounds expensive. Imagine Mojo wants to use Enzyme: it will have to implement the interfaces, but it doesn't work to carry the arith dependency because of that.

So I'm currently thinking about some centralized two-stage mechanism: First one must run "EnzymeInit" that registers all promises wrt upstream ops. Then one can run other registrations/loads as usual, but they would fail if "EnzymeInit" hasn't been called before. Basically forcing to register promises wrt upstream short of which the project is unusuable. Ideally registering these promises doesn't create a big dependency somehow...
